source/h2_stream.c

```diff
@@ -1114,6 +1152,36 @@ struct aws_h2err aws_h2_stream_on_decoder_end_stream(struct aws_h2_stream *strea
      * an actual frame type. It's a flag on DATA or HEADERS frames, and we
      * already checked the legality of those frames in their respective callbacks. */
 
+    if (stream->thread_data.content_length_received) {
+        if (stream->base.request_method != AWS_HTTP_METHOD_HEAD &&
+            stream->base.client_data->response_status != AWS_HTTP_STATUS_CODE_304_NOT_MODIFIED) {
```

Comment:
Huh. You're right that the HTTP/2 spec isn't required to enforce the body-less-ness of a HEAD or 304 response.

HTTP/1.1 is required to treat HEAD and 304 as having no body, because HTTP/1.1 is just going to treat the response as done after the headers section. It can't exactly validate it, because there's no "framing" just a raw stream of bytes.

I was going to say it seemed like an opportunity for request smuggling. Where HTTP/2 thinks it's 1 message, but if that's forwarded to HTTP/1.1 you could trick it into processing multiple messages...

But this only applies to responses. And I'm not sure *response* smuggling is a security concern? only *request* smuggling?